tests/trainer/test_trainer.py

```diff
@@ -1,247 +1,375 @@
+            algorithm = hparams(num_classes=2).initialize_object()
+        else:
+            algorithm = hparams().initialize_object()
+
+        return algorithm
+
+    @pytest.fixture(params=callback_registry.items(), ids=callback_registry.keys())
+    def callback(self, request, tmpdir, monkeypatch):
+        name, hparams = request.param
+
+        if name == 'benchmarker':
+            pytest.skip('Deprecated callback.')
+
+        # create callback
+        if name == 'run_directory_uploader':
+            pytest.importorskip('libcloud', reason='libcloud required.')
+            monkeypatch.setenv("KEY_ENVIRON", str(tmpdir))
+
+            callback = hparams(
+                provider='local',
+                container='.',
+                key_environ="KEY_ENVIRON",
+            ).initialize_object()
+        else:
+            callback = hparams().initialize_object()
+
+        return callback
+
+    @pytest.fixture(params=logger_registry.items(), ids=logger_registry.keys())
+    def logger(self, request):
+
+        name, hparams = request.param
+        skip_list = {
+            'wandb': 'Requires wandb account.',
+            'mosaicml': 'Not supported',
+        }
+        if name in skip_list:
+            pytest.skip(skip_list[name])
+
+        return hparams().initialize_object()
+
+    """
+    Tests that training completes.
+    """
+
+    def test_algorithms(self, config, algorithm):
+        config['algorithms'] = [algorithm]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    @pytest.mark.timeout(10)
+    def test_callbacks(self, config, callback):
+        config['callbacks'] = [callback]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    def test_loggers(self, config, logger):
+        config['loggers'] = [logger]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    """
+    Tests that training with multiple fits complete.
+    Note: future functional tests should test for
+    idempotency (e.g functionally)
+    """
+
+    def test_algorithms_multiple_calls(self, config, algorithm):
+        if isinstance(algorithm, LayerFreezing):
+            pytest.xfail("Known idempotency issue.")
+        config['algorithms'] = [algorithm]
+        trainer = Trainer(**config)
+        self._test_multiple_fits(trainer)
+
+    def test_callbacks_multiple_calls(self, config, callback):
+        if isinstance(callback, RunDirectoryUploader):
+            pytest.xfail("Known idempotency issue.")
+        config['callbacks'] = [callback]
+        trainer = Trainer(**config)
+        self._test_multiple_fits(trainer)
+
+    def test_loggers_multiple_calls(self, config, logger):
+        if isinstance(logger, FileLogger):
+            pytest.xfail("Known idempotency issue.")
+        config['loggers'] = [logger]
+        trainer = Trainer(**config)
+        self._test_multiple_fits(trainer)
+
+    def _test_multiple_fits(self, trainer):
+        trainer.fit()
```

Comment:
I think the private method for the first call to fit should be used so callbacks aren't closed.
(Once we officially support multiple calls to fit, then we won't need to use the private method)

Reply:
If we aim to support multiple calls to fit, better for the test to support exactly code path and `xfail` to highlight specific issues. Otherwise, users may read this test and think that calling the private method is the recommended approach for multiple calls.

Reply:
OK, makes sense.